src/hooks/UseGraph/Graph.ts

```diff
@@ -0,0 +1,1147 @@
+   *
+   * @type {Set<TData, TDimensions>}
+   */
+  set: Set<TData, TDimensions>
+  /**
+   * Sets the graph at certain coordinates
+   *
+   * @type {SetAtCoordinates<TData, TDimensions>}
+   */
+  setAtCoordinates: SetAtCoordinates<TData, TDimensions>
+  /**
+   * Sets the graph at a vertex
+   *
+   * @type {SetVertex<TData, TDimensions>}
+   */
+  setVertex: SetVertex<TData, TDimensions>
+
+  /**
+   * Transforms the graph
+   *
+   * @type {Update<TData, TDimensions>}
+   */
+  update: Update<TData, TDimensions>
+  /**
+   * Transforms the graph at certain coordinates
+   *
+   * @type {UpdateAtCoordinates<TData, TDimensions>}
+   */
+  updateAtCoordinates: UpdateAtCoordinates<TData, TDimensions>
+  /**
+   * Sets the graph at a vertex
+   *
+   * @type {UpdateVertex<TData, TDimensions>}
+   */
+  updateVertex: UpdateVertex<TData, TDimensions>
+
+  /**
+   * Empties the graph
+   *
+   * @type {Clear}
+   */
+  clear: Clear
+  /**
+   * Removes the coordinates and everything below them from the graph
+   *
+   * @type {PruneAtCoordinates<TDimensions>}
+   */
+  pruneAtCoordinates: PruneAtCoordinates<TDimensions>
+  /**
+   * Removes a vertex from the graph
+   *
+   * @type {PruneVertex<TDimensions>}
+   */
+  pruneVertex: PruneVertex<TDimensions>
+
+  /**
+   * Returns a transformed graph
+   *
+   * @type {Map<TData, TDimensions>}
+   */
+  map: Map<TData, TDimensions>
+  /**
+   * Returns transformed data at certain coordinates
+   *
+   * @type {MapAtCoordinates<TData, TDimensions>}
+   */
+  mapAtCoordinates: MapAtCoordinates<TData, TDimensions>
+  /**
+   * Returns transformed data at a vertex
+   *
+   * @type {MapVertex<TData, TDimensions>}
+   */
+  mapVertex: MapVertex<TData, TDimensions>
+
+  /**
+   * Executes logic for each vertex of the graph
+   *
+   * @type {ForEachVertex<TData, TDimensions>}
+   */
+  forEachVertex: ForEachVertex<TData, TDimensions>
+
+  /**
+   * Returns `true` if calling the `callback` on any vertex returns `true`
+   *
+   * @type {SomeVertex<TData, TDimensions>}
+   */
+  someVertex: SomeVertex<TData, TDimensions>
+
+  /**
+   * Transforms all vertices in the graph
+   *
+   * @type {UpdateAllVertices<TData, TDimensions>}
+   */
+  updateAllVertices: UpdateAllVertices<TData, TDimensions>
+
+  /**
+   * Sets all vertices in the graph to a new value
+   *
+   * @type {SetAllVertices<TData>}
+   */
+  setAllVertices: SetAllVertices<TData>
+
+  /**
+   * Returns a graph where all `TData` vertices have been transformed to `TResult`s
+   *
+   * @type {MapAllVertices<TData, TDimensions>}
+   */
+  mapAllVertices: MapAllVertices<TData, TDimensions>
+}
+
+/**
+ * A data structure capable of containing a graph of any depth
+ *
+ * @export
+ * @class Graph
+ * @typedef {Graph}
+ * @template TData
+ * @template {number} [TDimensions=0]
+ * @implements {IGraph<TData, TDimensions>}
+ */
+export class Graph<TData, TDimensions extends number = 0> implements IGraph<TData, TDimensions> {
+  // #region Graph attributes
```

Comment:
I'm not sure we want to commit these regions

Reply:
Should we include a region parser VSCode extension in the file for recommended extensions? Unless it's just a setting. I can't get my VSCode to respect the regions :(

Reply:
Oh weird it just worked for me. Maybe I have another extension that makes it work by accident. I added a recommendation for a region folding extension, lmk if it still doesn't work with that and we can re-evaluate.

Reply:
Got it working without the extension!